v0.8.2 release blog (#97)
kuzudb/kuzu-blog: src/content/post/2025-02-19-kuzu-v-0.8.1.md

---
slug: "kuzu-0.8.1-release"
title: "Kùzu 0.8.1 Release"
description: "Release announcement for Kùzu 0.8.1"
pubDate: "Feb 19 2025"
heroImage: "/img/default.png"
categories: ["release"]
authors: ["team"]
tags: ["cypher"]
---

We're just a couple months into 2025, and we are happy to announce a new minor release: v0.8.1. This release is feature-packed, warranting its own blog post. One of the highlights is the introduction of the `unity_catalog` extension, which allows you to scan/copy from Delta Lake tables managed by [Unity Catalog](https://www.unitycatalog.io/).

> We've also improved our existing extensions. For those of you on Google Cloud, we have some exciting news! We now support scanning/copying from/writing to files hosted on [Google Cloud Storage](https://cloud.google.com/storage)(GCS) filesystem. This update leverages our existing `httpfs` extension. Another useful new feature is that our CLI now explicitly excludes confidential information such as `S3 access keys` from being stored in the command history file. This helps prevent accidental leakage of sensitive data into your command line history and ensures your credentials remain secure.

Reviewer: ```suggestion
We've also improved our existing extensions. For those of you on Google Cloud, we have some exciting news! We now support scanning/copying from/writing to files hosted on [Google Cloud Storage](https://cloud.google.com/storage)(GCS) filesystem. This update leverages our existing `httpfs` extension. Another useful new feature is that our CLI now explicitly excludes confidential information such as `S3 access keys` from being stored in the command history file. This helps prevent accidental leakage of sensitive data into your command line history and ensures your credentials remain secure.
```
```suggestion
We've also improved our existing extensions. For those of you on Google Cloud, we have some exciting news! We now support scanning/copying from/writing to files hosted on a [Google Cloud Storage](https://cloud.google.com/storage) (GCS) filesystem. This update leverages our existing `httpfs` extension. We also added a usability enhancement to our CLI, which now explicitly excludes confidential information such as `S3 access keys` from being stored in the command history file. This helps prevent accidental leakage of sensitive data into your command line history and ensures your credentials remain secure.
```

From a performance perspective, we’ve significantly improved our execution of distinct aggregation queries via a new parallel distinct hash aggregation mechanism.
In the following sections, we dive into details.

## New features

### Unity Catalog extension

Unity Catalog, developed by Databricks, is an open data governance solution that allows you to manage your data catalogs, data access, and data sharing. It provides a unified interface for managing data across different data storage systems, including databases, data lakes, and data warehouses.
For Kùzu users who are entrenched in the data lake ecosystem, we've introduced an experimental `unity_catalog` extension to enable scanning/copying from Delta Lake tables managed by [Unity Catalog](https://www.unitycatalog.io/]).

A quickstart guide on how to setup a local unity catalog server can be found: [there](https://docs.unitycatalog.io/quickstart)

To attach a unity catalog, you have to first install and load the unity_catalog extension:
```
INSTALL unity_catalog;
LOAD EXTENSION unity_catalog;
```

> Similar to attach to a relational database, unity catalog can be attached using the `ATTACH` statement. The example below attaches to the catalog named `unity`:

Reviewer: ```suggestion
Similar to attaching to a relational database, unity catalog can be attached using the `ATTACH` statement. The example below attaches to the catalog named `unity`:
```

```
ATTACH 'unity' AS unity (dbtype uc_catalog)
```

> Once the `unity` catalog is attached, you can scan from delta tables hosted by the `unity` catalog under the `default` schema as follow:

Reviewer: ```suggestion
Once the `unity` catalog is attached, you can scan from delta tables hosted by the `unity` catalog under the `default` schema as follows:
```

```cypher
// Scan the numbers table
LOAD FROM unity.default.numbers
RETURN *
```

One important use case of the Unity Catalog extension is to facilitate seamless data transfer from tables in Unity Catalog to Kùzu.
```cypher
// Migrate data from the unity.default.numbers table to the `numbers` table in Kùzu
CREATE NODE TABLE numbers (id INT32, score DOUBLE , PRIMARY KEY(id));
COPY numbers FROM unity.numbers;
```

### Google Cloud Storage (GCS) support

> Our `httpfs` extension now supports working with files on [Google Cloud Storage](https://cloud.google.com/storage). Any operation that we earlier supported with Amazon S3 will now also work with GCS. This includes:

Reviewer: ```suggestion
We're happy to announce that our `httpfs` extension now supports working with files on [Google Cloud Storage](https://cloud.google.com/storage). Any operation that we earlier supported with Amazon S3 will now also work with GCS. This includes:
```

- Scanning (`COPY/LOAD FROM`) from files hosted on GCS
- Writing to (`COPY TO`) files hosted on GCS
- Attaching to a read-only kuzu database hosted on GCS

To work with GCS, you must configure kuzu to use your `GCS HMAC keys`, per [the official docs](https://cloud.google.com/storage/docs/authentication/hmackeys):

```cypher
CALL gcs_access_key_id='${GCS_ACCESS_KEY_ID}';
CALL gcs_secret_access_key='${GCS_SECRET_ACCESS_KEY}'
```

After configuring your credentials, you can access files on GCS by using urls in the format `gcs://${BUCKET_NAME}/${PATH_TO_FILE_IN_BUCKET}`:

```cypher
// scan from GCS
LOAD FROM 'gs://kuzudb-test/user.csv' RETURN *; 
// upload to GCS
COPY (MATCH (p:Person) RETURN p) TO 'gcs://kuzudb-test/person.csv' (header=true); 
// attach remote DB in GCS
ATTACH 'gcs://kuzudb-test/tinysnb/' AS tinysnb (dbtype kuzu); 
```

You can check out the documentation for our GCS feature [here](https://dev-docs.kuzudb.com/extensions/httpfs/#gcs-file-system).

### Stopwords customization in Full-text Search (FTS)

> Stopwords in full-text search (FTS) are commonly occurring words that are excluded from indexing and query processing to improve efficiency and relevance. Words like "the," "and," "is," or "in" (language-specific), which are deemed non-essential for search relevance, are ignored during the index creation step.

Reviewer: ```suggestion
Stopwords in full-text search (FTS) are commonly occurring words that are excluded from indexing and query processing to improve efficiency and relevance. Common words like "the", "and", "is", "in", etc. (which can be language-specific), are deemed non-essential for search relevance and are ignored during the index creation step.
```

By default, Kùzu uses a pre-defined list of common english stopwords. However it's perfectly plausible that you may need to customize the stopwords list to be more useful for your domain and use cases. In v0.8.1, a custom stopword list can be given as an optional parameter when an FTS index is built as follows:

```cypher
CALL CREATE_FTS_INDEX(
    'Book',   // Table name
    'book_index',   // Index name
    ['abstract', 'title'],   // Properties to build FTS index on
    stemmer := 'porter',   // Stemmer to use (optional)
    stopwords := 'https://stopwords/porter.csv' // Configure customized stopwords list (optional)
); 
```

The `stopwords` parameter can either be:

> 1. A file path to a single string column `CSV`/`PARQUET` file with each stopword on a separate line.

Reviewer: ```suggestion
1. A file path to a single string column in a CSV or Parquet file, with each stopword on a separate line.
```

> 2. A node table with a single string column with each stopword on a separate row.

Reviewer: ```suggestion
2. A node table with a single string column, with each stopword on a separate row.
```
